libclc/utils/libclc-remangler/LibclcRemangler.cpp

```diff
@@ -536,157 +778,204 @@ class TargetTypeReplacements {
+                     const itanium_demangle::Node *FunctionTree,
+                     SmallDenseMap<const char *, const char *> TypeReplacements,
+                     bool CloneeTypeReplacement = false) {
+    Remangler ATR{ContextAST, FunctionTree, TypeReplacements};
 
-bool createClones(Module *M, std::string originalMangledName,
-                  std::string remangledName,
-                  const itanium_demangle::Node *functionTree,
-                  TargetTypeReplacements replacements) {
-  // create clone of original function
-  if (!createCloneFromMap(M, originalMangledName, functionTree,
-                          replacements.getCloneTypeReplacements(),
-                          /* CloneeTypeReplacement= */ true))
-    return false;
-
-  // create clone of remangled function
-  if (!createCloneFromMap(M, remangledName, functionTree,
-                          replacements.getRemangledCloneTypeReplacements()))
-    return false;
-
-  return true;
-}
+    std::string const RemangledName = ATR.remangle();
 
-bool remangleFunction(Function &func, Module *M,
-                      TargetTypeReplacements replacements) {
-  if (!func.getName().startswith("_Z"))
-    return true;
+    if (ATR.hasFailed())
+      return false;
 
-  std::string MangledName = func.getName().str();
-  Demangler D{MangledName.c_str(), MangledName.c_str() + MangledName.length()};
-  const itanium_demangle::Node *FunctionTree = D.parse();
-  if (FunctionTree == nullptr) {
-    errs() << "Unable to demangle name: " << MangledName << '\n';
-    return false;
-  }
+    // Name has not changed from the original name.
+    if (RemangledName == OriginalName)
+      return true;
 
-  // Try to change the parameter types in the function name using the mappings.
-  Remangler R{FunctionTree, replacements.getParameterTypeReplacements()};
-  std::string RemangledName = R.remangle();
+    StringRef CloneName, CloneeName;
+    if (CloneeTypeReplacement) {
+      CloneName = OriginalName;
+      CloneeName = RemangledName;
+    } else {
+      CloneName = RemangledName;
+      CloneeName = OriginalName;
+    }
 
-  if (R.hasFailed())
-    return false;
+    Function *Clonee = M->getFunction(CloneeName);
+    if (Clonee) {
+      ValueToValueMapTy Dummy;
+      Function *NewF = CloneFunction(Clonee, Dummy);
+      NewF->setName(std::string(CloneName));
+    } else if (Verbose) {
+      std::cout << "Could not create copy " << CloneName.data() << " : missing "
+                << CloneeName.data() << std::endl;
+    }
 
-  if (RemangledName != MangledName) {
-    if (Verbose) {
-      std::cout << "Mangling changed:"
-                << "\n"
-                << "Original:  " << MangledName << "\n"
-                << "New:       " << RemangledName << "\n"
-                << std::endl;
+    return true;
+  }
+
+  bool remangleFunction(Function &Func, llvm::Module *M) {
+    if (!Func.getName().startswith("_Z"))
+      return true;

```

Comment:
Should this return false instead?

Reply:
I don't think so, the bool here is used as a success/failure notion, and I'd say that a state where we don't need to remangle and can early exit from this function matches the success.